ethereum/contractBackend.go

```diff
@@ -0,0 +1,97 @@
+package ethereum
+
+import (
+	"context"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts"
+	"github.com/ethereum/go-ethereum/accounts/keystore"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/pkg/errors"
+
+	ethchannel "perun.network/go-perun/backend/ethereum/channel"
+	ethwallet "perun.network/go-perun/backend/ethereum/wallet"
+	"perun.network/go-perun/channel"
+	"perun.network/go-perun/wallet"
+)
+
+// ContractBackend provides ethereum specific contract backend functionality.
+type ContractBackend struct {
+	// cb is the instance of contract backend that will be used for all onchain communications.
+	cb *ethchannel.ContractBackend
```

Comment:
Can you embed or typedef this?

Reply:
I use `ethchannel.contractBackend` as a private member instead of embedding because the `ContractBackend` struct in this package implements the `ContractBackend` interface defined in `client/client.go`. And, I do not want this struct to expose the methods of the `ethchannel.ContractBackend`. The purpose of this interface is to make all the ethereum specific functions accessible using the abstract types defined in the core go-perun sdk.

Packages in `go-ethereum`, that are required in this project are licensed under LGPL v3 and it is one of the objectives to have a dst-go binary that does not contain any code licensed under LGPL v3.

By this arrangement all imports from `go-perun/backend/ethereum` and `go-ethereum` are restricted only to ethereum package. While the dst-go binary may still contain LGPL code as the `dst-go/ethereum` is statically linked with other packages, it is always possible (since we have proper abstractions defined) to covert the static linking to dynamic linking by using the `plugin` package in go, whenever we require a `dst-go` binary that is not containing any LGPL code.

Reply:
Im not familiar with the licence stuff, @sebastianst can have a look when he has time.